README.md
# SKC
The [Steven Kruijswijk Coefficient](https://www.bicycling.com/nl/stijl/a26851320/de-steven-kruijswijk-coefficient/) is a number that represents the ratio of ridden time compared to the elapsed time. It was first mentioned by Lauren ten Dam during his podcast series [Live Slow, Ride Fast](https://www.liveslowridefast.com/).

## Installation
The extension is available in the add-on store for Firefox: https://addons.mozilla.org/en-US/firefox/addon/skc/

For Chrome the extension is not yet available in the Chrome Store.

### Manual installation
With every release a zip file with the necessary files is added to the release. The releases are listed in the [Releases](https://github.com/timoschinkel/skc/releases) section of Github. The file should be called `skc-[version].zip`

Using this zip you can install the extension as a temporary extension in both Firefox and Chrome.

For Firefox:
- Download and unpack the zip file
- Navigate to [about:debugging#/runtime/this-firefox](about:debugging#/runtime/this-firefox)
- Chose "Load temporary add-on" and select the `manifest.json` file

For Chrome:
- Download and unpack the zip file
- Navigate to [chrome://extensions/](chrome://extensions/)
- Enable "Developer mode" on the top right
- Select "Load unpacked" and select the folder where the extension is located

## What is this?
A browser extension that adds the Steven Kruijswijk Coefficient to your Strava rides. For now this extension only works with Strava. With enough demand I might look into making this extension work with other services similar to Strava as well.

## Why did I build this?
Basically; Why not? As a listener of the Live Slow, Ride Fast podcast I learned of the Steven Kruijswijk Coefficient. A number of tools like [Bicycling's calculator](https://tools.bicycling.nl/steven-kruijswijk-coefficient) (Dutch) and [What Is My SKC](https://whatismyskc.com/) have been made available. But these tools either require me to copy and paste times or to associate my Strava account with a separate website. I am a lazy man, I want to see my SKC in my Strava feed. 

## The calculated coefficient is wrong
The calculation used is the same as is used by [Bicyling's calculator](https://tools.bicycling.nl/steven-kruijswijk-coefficient). If the coefficient diverges from the outcome giving by that calculator please create [an issue](https://github.com/timoschinkel/skc/issues) on this repository with a public activity where the coefficient is wrong. This information is needed to reproduce any issues.

## The extension does not work for me!
I guess that can happen. The extension reads the moving time and elapsed time from the html of Strava and tries to inject the calculated SKC into that same html. It reuses existing utility classes of Strava. If Strava decides to change their html structure the extension might not work anymore. If that is the case please create [an issue](https://github.com/timoschinkel/skc/issues) on this repository with a public activity where the extension fails. This information is needed to reproduce any issues.

Please keep in mind that Steven Kruijswijk is a cyclist. That means that the extension is built to work with cycling rides. Any other activity types such as running activities might not work as Strava uses a different html structure for those activities.

## Can I help?
Yes you can. If you have an idea, suggestion or complaint you are free to open [an issue](https://github.com/timoschinkel/skc/issues) and you can also create [a pull request](https://github.com/timoschinkel/skc/pulls) to solve the issue. 

### Setting up the development environment
To develop the extension [NodeJS](https://nodejs.org/en/) and [NPM](https://www.npmjs.com/) are required. The codebase is tested on NodeJS 14.

After cloning the repository you need to install the dependencies via `npm ci`. Building the extension - so that you can load it as temporary extension on your browser - run `npm run build`. Making the extension available in your browser requires the same steps as described in the "Manual installation" section.


## Disclaimer
I am in no way affiliated with Steven Kruijswijk - or any of his current or previous cycling teams, Bicyling magazine, Laurens ten Dam or Live Slow, Ride Fast. I am just a developer that likes to ride my bike. I have built this extension because I thought it would be fun. Because I believe in sharing and open source I have made this extension available for everyone to use. That does mean that this extension is offered as is, without any guarantees.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-03-12
The initial version of the extension!

src/skc.js
export const STEVIE_STYLE = 'Stevie style';
export const ON_PAR = 'On par';
export const HTFU = 'HTFU';

/**
 * @param {Number} moving_time 
 * @param {Number} elapsed_time 
 * @returns {Number}
 */
export const calculate_skc = (moving_time, elapsed_time) => 
    Math.round(((elapsed_time - moving_time) / moving_time) * 1000) / 10;

const moving_time_per_hour = (hours) =>
    3600 * hours;

/**
 * @param {Number} moving_time 
 * @param {Number} skc
 * @returns {Object}
 */
export const grade_skc = (moving_time, skc) => {
    // Source for this mapping the Bicycling magazine. Translations are my own.
    // https://tools.bicycling.nl/steven-kruijswijk-coefficient
    const mapping = [
        { moving_time_floor: moving_time_per_hour(1), stevie_style_ceiling: -1, on_par_ceiling: 0, text: __('Did you really stop on a one hour ride?') },
        { moving_time_floor: moving_time_per_hour(2), stevie_style_ceiling: -1, on_par_ceiling: 5, text: __('5 minute break max, we all need to stop for a traffic light.') },
        { moving_time_floor: moving_time_per_hour(3), stevie_style_ceiling: 5, on_par_ceiling: 10, text: __('Three hour ride? A coffee break is allowed, but 20 minutes max!') },
        { moving_time_floor: moving_time_per_hour(4), stevie_style_ceiling: 6, on_par_ceiling: 8 , text: __('Four hours? A coffee break is allowed, but that\'s it. 20 minutes. Do it.') },
        { moving_time_floor: moving_time_per_hour(5), stevie_style_ceiling: 6, on_par_ceiling: 8, text: __('With a five hour ride: coffee break and candy bar and fizzy drink is allowed. 25 minutes without movement max.') },
        { moving_time_floor: moving_time_per_hour(6), stevie_style_ceiling: 6, on_par_ceiling: 8, text: __('Six hour ride? Coffee break and a candy bar and fizzy drink stop is allowed. And a toilet break. 30 minutes. Full throttle.') },
        { moving_time_floor: moving_time_per_hour(7), stevie_style_ceiling: 7, on_par_ceiling: 10, text: __('Seven hours. Okay, okay. Coffee break and a candy bar and fizzy drink stop allowed. And a toilet break. Or two. 45 minutes.') },
        { moving_time_floor: moving_time_per_hour(8), stevie_style_ceiling: 8, on_par_ceiling: 11, text: __('Eight hours? Nice job. You\'ll get 50 minutes of "me time". Enjoy.') },
        { moving_time_floor: moving_time_per_hour(9), stevie_style_ceiling: 8, on_par_ceiling: 11, text: __('Nine hour training? Kudo\'s guaranteed. 60 minutes to be on par with SKC!') },
        { moving_time_floor: moving_time_per_hour(10), stevie_style_ceiling: 10, on_par_ceiling: 13, text: __('Sick fuck. (Take your time.)') },
    ];

    const eligible = mapping.filter(({ moving_time_floor }) => moving_time_floor <= moving_time);
    const rank = eligible[eligible.length - 1];

    if (skc <= rank.stevie_style_ceiling) {
        return { grade: STEVIE_STYLE, text: rank.text };
    } else if (skc <= rank.on_par_ceiling) {
        return { grade: ON_PAR, text: rank.text };
    }

    return { grade: HTFU, text: rank.text };
}

src/i18n.js
const translations = {
    'nl-NL': {
        'Steven Kruijswijk Coefficient': 'Steven Kruijswijk Coëfficiënt',
        'Did you really stop on a one hour ride?': 'Gestopt op een ritje van een uur? Twintig zweepslagen van Stevie zelf.',
        '5 minute break max, we all need to stop for a traffic light.': 'Vijf minuten stilstaan max.  Vooruit dan. We komen allemaal wel eens een verkeerslichtje tegen waar je echt naar moet luisteren.',
        'Three hour ride? A coffee break is allowed, but 20 minutes max!': 'Drie uur fietsen? Koffie-break allowed! Maar rap wat. 20 min max!',
        'Four hours? A coffee break is allowed, but that\'s it. 20 minutes. Do it.': 'Vier uur? Koffiebreak mag, maar that\'s it. 20 min. Do it.',
        'With a five hour ride: coffee break and candy bar and fizzy drink is allowed. 25 minutes without movement max.': 'Bij vijf uur fietsen: koffie-break en snicker-cola break toegestaan. 25 min stilstaan max.',
        'Six hour ride? Coffee break and a candy bar and fizzy drink stop is allowed. And a toilet break. 30 minutes. Full throttle.': 'Zes uur rijden? Koffie- en snicker-cola break mag. Plus ff pissen. 30 min. Gas erop.',
        'Seven hours. Okay, okay. Coffee break and a candy bar and fizzy drink stop allowed. And a toilet break. Or two. 45 minutes.': 'Zeven uur. Ok, ok. Koffie- en snicker-cola break mag. Plus ff pissen. Of twee keer. Vooruit. 45min.',
        'Eight hours? Nice job. You\'ll get 50 minutes of "me time". Enjoy.': 'Acht uur? Lekker sjef. 50 minuten me time krijg je. Geniet er van.',
        'Nine hour training? Kudo\'s guaranteed. 60 minutes to be on par with SKC!': 'Negen uur trainen? Kudo’s sowieso. 60 minuten voor een on par SKC!',
        'Sick fuck. (Take your time.)': 'Sick fuck. (Take your time.)',
    }
}

/**
 * 
 * @param {String} translation_string 
 * @returns {String}
 */
export const __ = (translation_string) =>
    translations?.[language]?.[translation_string] || translation_string;

src/content_scripts/strava.js
'use strict';

import { STEVIE_STYLE, ON_PAR, HTFU, calculate_skc, grade_skc } from './../skc';
import { __ } from './../i18n';
import { duration_to_seconds } from './../time';

/** @type {String|undefined} */
const language = document.querySelector('html[lang]')?.getAttribute('lang');

const inject = (skc, { grade, text }) => {
    const activity_stats = document.querySelector('.activity-stats');
    if (activity_stats) {

        let color = 'inherit';
        if (grade == STEVIE_STYLE) {
            color = '#36c597';
        }
        if (grade == HTFU) {
            color = '#b40312';
        }
        
        const localized_skc = Intl.NumberFormat(language).format(skc);
        const localized_title = __('Steven Kruijswijk Coefficient');

        activity_stats.innerHTML += `<div class="section skc">
            <div class="skc-stats row">
                <div class="skc-number spans8">
                    <div style="font-size: 28px; color: ${color};" title="${grade}">${localized_skc}</div>
                    <div>${localized_title}</div>
                </div>
                <div class="skc-grade spans8">${text}</div>
            </div>
        </div>`;
    }
}

const get_moving_time = () => {
    const time = document.querySelector('[data-glossary-term="definition-moving-time"]')
        ?.closest('li')
        ?.querySelector('strong')
        ?.innerText;
    
    return time ? duration_to_seconds(time) : undefined;
}

const get_elapsed_time = () => {
    const time = document.querySelector('[data-glossary-term="definition-elapsed-time"]')
        ?.closest('tr')
        ?.querySelector('td')
        ?.innerText;
    
    return time ? duration_to_seconds(time) : undefined;
}

const moving_time = get_moving_time();
const elapsed_time = get_elapsed_time();

if (moving_time && elapsed_time) {
    const skc = calculate_skc(moving_time, elapsed_time);
    const grade = grade_skc(moving_time, skc);

    inject(skc, grade);
}


src/inject_into_strava.js
'use strict';

const STEVIE_STYLE = 'Stevie style';
const ON_PAR = 'On par';
const HTFU = 'HTFU';

/** @type {String|undefined} */
const language = document.querySelector('html[lang]')?.getAttribute('lang');

const translations = {
    'nl-NL': {
        'Steven Kruijswijk Coefficient': 'Steven Kruijswijk Coëfficiënt',
        'Did you really stop on a one hour ride?': 'Gestopt op een ritje van een uur? Twintig zweepslagen van Stevie zelf.',
        '5 minute break max, we all need to stop for a traffic light.': 'Vijf minuten stilstaan max.  Vooruit dan. We komen allemaal wel eens een verkeerslichtje tegen waar je echt naar moet luisteren.',
        'Three hour ride? A coffee break is allowed, but 20 minutes max!': 'Drie uur fietsen? Koffie-break allowed! Maar rap wat. 20 min max!',
        'Four hours? A coffee break is allowed, but that\'s it. 20 minutes. Do it.': 'Vier uur? Koffiebreak mag, maar that\'s it. 20 min. Do it.',
        'With a five hour ride: coffee break and candy bar and fizzy drink is allowed. 25 minutes without movement max.': 'Bij vijf uur fietsen: koffie-break en snicker-cola break toegestaan. 25 min stilstaan max.',
        'Six hour ride? Coffee break and a candy bar and fizzy drink stop is allowed. And a toilet break. 30 minutes. Full throttle.': 'Zes uur rijden? Koffie- en snicker-cola break mag. Plus ff pissen. 30 min. Gas erop.',
        'Seven hours. Okay, okay. Coffee break and a candy bar and fizzy drink stop allowed. And a toilet break. Or two. 45 minutes.': 'Zeven uur. Ok, ok. Koffie- en snicker-cola break mag. Plus ff pissen. Of twee keer. Vooruit. 45min.',
        'Eight hours? Nice job. You\'ll get 50 minutes of "me time". Enjoy.': 'Acht uur? Lekker sjef. 50 minuten me time krijg je. Geniet er van.',
        'Nine hour training? Kudo\'s guaranteed. 60 minutes to be on par with SKC!': 'Negen uur trainen? Kudo’s sowieso. 60 minuten voor een on par SKC!',
        'Sick fuck. (Take your time.)': 'Sick fuck. (Take your time.)',
    }
}

/**
 * 
 * @param {String} translation_string 
 * @returns {String}
 */
const __ = (translation_string) =>
    translations?.[language]?.[translation_string] || translation_string;

/**
 * @typedef {Object} SKC
 * @property {Number} skc - The calculated Steven Kruijswijk Coefficient
 * @property {String} grade - Stevie Style, on par or HTFU
 * @property {String} text - Textual comment on your skc score
 */

/**
 * @param {Number} moving_time
 * @param {Number} elapsed_time
 * @param {SKC}
 */
const calculate_skc = (moving_time, elapsed_time) => {
    // Source for this mapping the Bicycling magazine. Translations are my own.
    // https://tools.bicycling.nl/steven-kruijswijk-coefficient
    const mapping = [
        { hour: 1, stevie_style_ceiling: -1, on_par_ceiling: 0, text: __('Did you really stop on a one hour ride?') },
        { hour: 2, stevie_style_ceiling: -1, on_par_ceiling: 5, text: __('5 minute break max, we all need to stop for a traffic light.') },
        { hour: 3, stevie_style_ceiling: 5, on_par_ceiling: 10, text: __('Three hour ride? A coffee break is allowed, but 20 minutes max!') },
        { hour: 4, stevie_style_ceiling: 6, on_par_ceiling: 8 , text: __('Four hours? A coffee break is allowed, but that\'s it. 20 minutes. Do it.') },
        { hour: 5, stevie_style_ceiling: 6, on_par_ceiling: 8, text: __('With a five hour ride: coffee break and candy bar and fizzy drink is allowed. 25 minutes without movement max.') },
        { hour: 6, stevie_style_ceiling: 6, on_par_ceiling: 8, text: __('Six hour ride? Coffee break and a candy bar and fizzy drink stop is allowed. And a toilet break. 30 minutes. Full throttle.') },
        { hour: 7, stevie_style_ceiling: 7, on_par_ceiling: 10, text: __('Seven hours. Okay, okay. Coffee break and a candy bar and fizzy drink stop allowed. And a toilet break. Or two. 45 minutes.') },
        { hour: 8, stevie_style_ceiling: 8, on_par_ceiling: 11, text: __('Eight hours? Nice job. You\'ll get 50 minutes of "me time". Enjoy.') },
        { hour: 9, stevie_style_ceiling: 8, on_par_ceiling: 11, text: __('Nine hour training? Kudo\'s guaranteed. 60 minutes to be on par with SKC!') },
        { hour: 10, stevie_style_ceiling: 10, on_par_ceiling: 13, text: __('Sick fuck. (Take your time.)') },
    ];

    const skc = Math.round(((elapsed_time - moving_time) / moving_time) * 1000) / 10;
    const hour = Math.round(moving_time / 3600);

    const rank = mapping[Math.min(Math.max(0, hour - 1), mapping.length - 1)];

    if (skc <= rank.stevie_style_ceiling) {
        return { skc, grade: STEVIE_STYLE, text: rank.text };
    } else if (skc <= rank.on_par_ceiling) {
        return { skc, grade: ON_PAR, text: rank.text };
    }

    return { skc, grade: HTFU, text: rank.text };

}

const inject = ({skc, grade, text }) => {
    const activity_stats = document.querySelector('.activity-stats');
    if (activity_stats) {

        let color = 'inherit';
        if (grade == STEVIE_STYLE) {
            color = '#36c597';
        }
        if (grade == HTFU) {
            color = '#b40312';
        }
        
        const localized_skc = Intl.NumberFormat(language).format(skc);
        const localized_title = __('Steven Kruijswijk Coefficient');

        activity_stats.innerHTML += `<div class="section skc">
            <div class="skc-stats row">
                <div class="skc-number spans8">
                    <div style="font-size: 28px; color: ${color};" title="${grade}">${localized_skc}</div>
                    <div>${localized_title}</div>
                </div>
                <div class="skc-grade spans8">${text}</div>
            </div>
        </div>`;
    }
}

const duration_to_seconds = (time) => {
    const parts = time.split(':').reverse();
    
    return parseInt(parts[0])
        + (parts.length > 1 ? (parseInt(parts[1]) * 60) : 0)
        + (parts.length > 2 ? (parseInt(parts[2]) * 3600) : 0);

}

const get_moving_time = () => {
    const time = document.querySelector('[data-glossary-term="definition-moving-time"]')
        ?.closest('li')
        ?.querySelector('strong')
        ?.innerText;
    
    return time ? duration_to_seconds(time) : undefined;
}

const get_elapsed_time = () => {
    const time = document.querySelector('[data-glossary-term="definition-elapsed-time"]')
        ?.closest('tr')
        ?.querySelector('td')
        ?.innerText;
    
    return time ? duration_to_seconds(time) : undefined;
}

const moving_time = get_moving_time();
const elapsed_time = get_elapsed_time();

if (moving_time && elapsed_time) {
    const skc = calculate_skc(moving_time, elapsed_time);

    inject(skc);
}
